### PR TITLE
Faceted filtering.

### LINKED ISSUE #158: Search: Facets for dynamic metadata
See also #120 
Continue to built on #144 by adding facets for dynamic metadata ((course, run and learningresource type)

loredynfacets

### PR DESCRIPTION
**Pull request by ShawnMilo**
Show counts of courses, runs, and item types, and allow filtering by them.

Closes #158.

All work in this branch is based on [this Haystack documentation](http://django-haystack.readthedocs.org/en/latest/faceting.html).

### CODE REVIEW

**Comment by pdpinch** [comment]
I got it working locally and on https://lore-branch-test.herokuapp.com

Is there a reason you're not using the styling from http://pureui.com/c/mit/lore/build001/ (not the checkboxes, just the styling)? @pwilkins told me had a template for it from a while back, if that helps.

**Comment by pdpinch** [comment]
Also, I'm noticing that the pagination links lose the search query. 

For example, if I'm looking at https://lore-branch-test.herokuapp.com/repositories/foo/?q=angular+momentum the next page link goes to https://lore-branch-test.herokuapp.com/repositories/foo/?page=2

**Comment by ShawnMilo** [comment]
Thanks, I didn't try pagination.

As for the UI, I couldn't find the Bootstrap styles for that so I just made it work for now. I'll check with @pwilkins to get the correct markup.

**Comment by ShawnMilo** [comment]
@pdpinch: Fixed pagination and rebased. As per our conversation on HipChat I'm holding off on any style changes until @Ferdi decides which need to be in this PR and which can be added later.

**Comment by Ferdi** [comment]
Great progress !

I'd try to merge this sooner and do ui changes on a different pr.
- Sent from my phone

**Comment by ShawnMilo** [comment]
Great, thanks. The taxonomy faceted search is the next ticket, so it could fit to make the UI changes there.

File: learningresources/models.py [hunk 1]
```diff
@@ -111,8 +106,7 @@ def save(self, *args, **kwargs):
             roles_update_repo(self, old_slug)
         return new_repository
 
-    class Meta:
-        # pylint: disable=missing-docstring
+    class Meta:  # pylint: disable=missing-docstring
```
Comment (giocalitri) [inline learningresources/models.py:109]: who did you changed this?

Comment (ShawnMilo) [inline learningresources/models.py:109]: For some reason it started failing tests in Python 3. Doing that fixed it.

File: ui/urls.py [hunk 2]
```diff
@@ -27,6 +28,14 @@
 )
 import cas.urls as cas_urls
 
+
+def get_sqs():
+    """Get custom SearchQuerySet for LORE"""
+    sqs = SearchQuerySet()
+    for facet in ("course", "run", "resource_type"):
```
Comment (giocalitri) [inline ui/urls.py:35]: this list should not be hardcoded here

Comment (ShawnMilo) [inline ui/urls.py:35]: Any suggestions on where to move it? It's not used anywhere else, and I don't see an obvious better place.

If it's only used in one place, it seems best to keep it near where it's used.

Comment (giocalitri) [inline ui/urls.py:35]: If they are used only here it's fine

File: learningresources/models.py (hunk 1 above)
Comment (giocalitri) [inline learningresources/models.py:109]: This is weird: it should work anyway

Comment (giocalitri) [inline learningresources/models.py:109]: I confirm, it does not give me problems: are you sure the problem was that one?

**Pull request merged**